src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc

```diff
@@ -0,0 +1,174 @@
+      if (iter == MapRequestToResponse.end()) {
+        VLOG(1) << absl::Substitute("Could not find any responses for frame type = $0",
+                                    request_match_key);
+        continue;
+      }
+      if (iter->second == 0xff) {
+        // Request without responses found
+        req_frame.consumed = true;
+        latest_resp_ts = req_frame.timestamp_ns + 1;
+        mqtt::Message dummy_resp;
+        entries.push_back({std::move(req_frame), std::move(dummy_resp)});
+        continue;
+      }
+      uint8_t response_match_value = iter->second;
+
+      // finding the first response frame with timestamp greater than request frame
+      auto first_timestamp_iter =
+          std::lower_bound(resp_deque.begin(), resp_deque.end(), req_frame.timestamp_ns,
+                           [](const mqtt::Message& message, const uint64_t ts) {
+                             return ts > message.timestamp_ns;
+                           });
+      if (first_timestamp_iter == resp_deque.end()) {
+        VLOG(1) << absl::Substitute("Could not find any responses after timestamp = $0",
+                                    req_frame.timestamp_ns);
+        continue;
+      }
+
+      // finding the first appropriate response frame with the desired control packet type and flags
+      auto response_frame_iter = std::find_if(
+          first_timestamp_iter, resp_deque.end(), [response_match_value](mqtt::Message& message) {
+            return ((message.control_packet_type << 4) | (static_cast<uint8_t>(message.dup) << 3) |
+                    (message.header_fields["qos"] << 1) | static_cast<uint8_t>(message.retain)) ==
+                   response_match_value;
+          });
+      if (response_frame_iter == resp_deque.end()) {
+        VLOG(1) << absl::Substitute(
+            "Could not find any responses with control packet type and flag = $0",
+            response_match_value);
+        continue;
+      }
+      mqtt::Message& resp_frame = *response_frame_iter;
+
+      req_frame.consumed = true;
+      resp_frame.consumed = true;
+      entries.push_back({std::move(req_frame), std::move(resp_frame)});
+    }
+
+    // clearing the req_deque and resp_deque
+    auto erase_until_iter = req_deque.begin();
+    auto iter = req_deque.begin();
+    while (iter != req_deque.end() && (iter->timestamp_ns < latest_resp_ts)) {
+      if (iter->consumed) {
+        ++erase_until_iter;
+      }
+      if (!iter->consumed && !(iter == req_deque.end() - 1) && ((erase_until_iter + 1)->consumed)) {
+        ++error_count;
+        ++erase_until_iter;
+      }
+      ++iter;
+    }
+    req_deque.erase(req_deque.begin(), erase_until_iter);
+  }
+
+  // iterate through all response dequeues to find out which ones haven't been consumed
+  for (auto& [packet_id, resp_deque] : *resp_frames) {
+    for (auto& resp : resp_deque) {
+      if (!resp.consumed) {
+        error_count++;
+      }
+    }
+    resp_deque.clear();
+  }
```

Comment:
For other protocols, we assume that a request will always be seen before a response. This allows us the ability to clear the response deque at the end of each stitching round. I'm not sure that assumption will hold for MQTT since the server and client can independently initiate `PUBLISH` requests (as I mentioned in another comment).

Can we add a comment here to make sure we test how server "client" frames work with this logic?

Reply:
This would be a follow up for the TODO comment for the case where PUBLISH ends up in the response deque, right? Another TODO comment to make sure we test the clearing of the response deque when such a thing occurs?

Reply:
Yeah that comment makes things clear.